examples/apps/src/ble_bas_peripheral.rs

```diff
@@ -106,20 +107,22 @@ async fn gatt_events_task<P: PacketPool>(server: &Server<'_>, conn: &GattConnect
     let reason = loop {
         match conn.next().await {
             GattConnectionEvent::Disconnected { reason } => break reason,
-            GattConnectionEvent::Gatt { event: Err(e) } => warn!("[gatt] error processing event: {:?}", e),
-            GattConnectionEvent::Gatt { event: Ok(event) } => {
-                match &event {
-                    GattEvent::Read(event) => {
-                        if event.handle() == level.handle {
+            GattConnectionEvent::Gatt { event } => {
+                match event.payload().incoming() {
+                    AttClient::Request(AttReq::Read { handle, .. })
+                    | AttClient::Request(AttReq::ReadBlob { handle, .. }) => {
+                        if handle == level.handle {
                             let value = server.get(&level);
-                            info!("[gatt] Read Event to Level Characteristic: {:?}", value);
+                            info!("[gatt] Read from Level Characteristic: {:?}", value);
                         }
                     }
-                    GattEvent::Write(event) => {
-                        if event.handle() == level.handle {
-                            info!("[gatt] Write Event to Level Characteristic: {:?}", event.data());
+                    AttClient::Request(AttReq::Write { handle, data })
+                    | AttClient::Command(AttCmd::Write { handle, data }) => {
+                        if handle == level.handle {
+                            info!("[gatt] Write to Level Characteristic: {:?}", data);
                         }
                     }
+                    _ => {}
```

Comment:
For most users, won't this API be a lot more verbose than strictly necessary? I keep thinking it would be nice to at least have some higher level Read/Write categorization like the original nrf-softdevice had.

Reply:
... i.e. restoring the `ReadEvent` and `WriteEvent`. OK but what do we do with the "other" events which are neither "single attribute read/read blob" (`ReadEvent`) nor "single attribute write/cmd write" (`WriteEvent`)? For example:
- Read by group or read by type requests
- Find by type requests
- Confirmations
- FindInfo
- etc. etc.

The whole change suggested here is around the idea that **all** ATT payloads are passed to the user in the form of a `GattEvent` so that the user has the right to accept/reject/custom-process **all**/any ATT payload.

Can you point a link to the original `nrf-softdevice` code so that I can check what they were doing?

Reply:
Note also that the old `ReadEvent`/`WriteEvent` logic was also destructuring the `GattData` in an irreversible way. Which made impossible for the user to do any custom processing on the incoming data. It was either accept or reject and that's it, but there was no way to "bypass" the attribute server (the new method `GattEvent::into_payload() -> GattData`), and then call `GattData::reply` or `GattData::send_unsolicited` because `GattData` itself was lost in the process of converting it to `ReadEvent` / `WriteEvent` / silently processing all others with the attribute server.

Reply:
Ditto - topic no longer applicable after the switch to the `GattEvent::Other` approach.